### PR TITLE
blog on customizing istio coredns domain names

### PR DESCRIPTION
**Pull request by christian-posta**

### CODE REVIEW

File: .spelling [hunk 1]
```diff
@@ -555,6 +558,7 @@ security_bulletin
 selinux
 serverless
 serviceaccount
+ServiceEntry
```
Comment (sdake) [inline .spelling:561]: please quote custom resources, technology names, etc.

File: .spelling [hunk 2]
```diff
@@ -366,6 +368,7 @@ Kops
 Krishnan
 Kuat
 Kube
+kube-dns
```
Comment (sdake) [inline .spelling:371]: this should be quoted in the document and removed from `.spelling.

File: .spelling [hunk 3]
```diff
@@ -162,6 +162,7 @@ configurability
 containerID
 ControlZ
 CoreDNS
+coredns
```
Comment (sdake) [inline .spelling:165]: this should be quoted in the documentation.

File: .spelling [hunk 4]
```diff
@@ -596,6 +602,8 @@ subnetworks
 subresource
 subresources
 substring
+suffix
```
Comment (sdake) [inline .spelling:605]: surprised this is not in the dictionary - are you certain?

File: content/en/blog/2020/custom-istio-dns-name/index.md [hunk 5]
```diff
@@ -0,0 +1,125 @@
+---
+title: "Customizing Istio DNS stub names"
```
Comment (sdake) [inline content/en/blog/2020/custom-istio-dns-name/index.md:2]: the header (title, subtitle, description) I believe are not checked for spelling errors. Run `make lint` to verify.

File: content/en/blog/2020/custom-istio-dns-name/index.md [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+subtitle: Changing or adding stub domains to make ServiceEntry hostnames routable
+description: Customizing global or region hostnames as declared in ServiceEntry(s) for routability.
+publishdate: 2020-05-05
+attribution: "Christian Posta (Solo.io)"
+keywords: [istio, dns, mTLS, coredns]
+---
+
+Istio can be run across multiple clusters and has the necessary configuration APIs to support [shared control plane](/docs/setup/install/multicluster/shared/) or a non-shared, [replicated control model](/docs/setup/install/multicluster/gateways). The motivation for deploying across multiple clusters has been covered excellently by fellow Istio contributor [Vadim Eisenberg](https://twitter.com/vadimeisenberg) in the [Multi-Mesh Deployments for Isolation and Boundary Protection](/blog/2019/isolated-clusters/) blog.
```
Comment (sdake) [inline content/en/blog/2020/custom-istio-dns-name/index.md:10]: Anything in `[]` I believe is not spell checked.

You will see an overall theme in this review - we wish to bound the `.spelling` file changes.

File: content/en/blog/2020/custom-istio-dns-name/index.md [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+
+One aspect of multi-cluster functionality enabling services to communicate with each other is service discovery. In an Istio replicated control-plane multi-cluster setup, you create [ServiceEntry](/docs/reference/config/networking/service-entry/)s in one cluster to point to services running in another cluster. For example:
```
Comment (sdake) [inline content/en/blog/2020/custom-istio-dns-name/index.md:12]: control plane has no dash.

Comment (sdake) [reply content/en/blog/2020/custom-istio-dns-name/index.md:12]: `you create [Service Entry](link) custom resources in one cluster to reference services running in another cluster.`

File: content/en/blog/2020/custom-istio-dns-name/index.md [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+      addresses:
+      - 240.0.0.2
+      endpoints:
+      - address: ${CLUSTER2_GW_ADDR}
+        ports:
+          http1: 15443 # Do not change this port value
+    EOF
+    {{< /text >}}
+
+This `ServiceEntry` creates an entry in the Istio service-discovery registry for `httpbin.bar.global` with endpoints that point to a different cluster's gateway. Without any other configuration changes, this hostname, `httpbin.bar.global` is not routable. In the docs for setting up multi-cluster replicated control plane, however, there is a [section for configuring the DNS stubbing](/docs/setup/install/multicluster/gateways/#setup-dns) for your Kubernetes cluster. When we set up the DNS stubbing, we tell `kube-dns` to delegate queries for `.global` to the Istio coredns component.
```
Comment (sdake) [inline content/en/blog/2020/custom-istio-dns-name/index.md:41]: point to a -> direct traffic to

Comment (sdake) [reply content/en/blog/2020/custom-istio-dns-name/index.md:41]: we->you

File: content/en/blog/2020/custom-istio-dns-name/index.md [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+
+But what if you don't want to use `.global` as the suffix? What if you have to keep some backward compatibility and the root domain name needs to be something custom? Or what if you want to be able to support multiple root domain suffixes? In the open-source project [Service Mesh Hub](https://github.com/solo-io/service-mesh-hub) that I work on, we had this exact scenario. Service Mesh Hub is used to [automate multi-cluster Istio service mesh deployments](https://docs.solo.io/service-mesh-hub/latest/getting_started/), and routing between clusters is a use case our users have. Here's how to configure it in Istio.
```
Comment (sdake) [inline content/en/blog/2020/custom-istio-dns-name/index.md:43]: Hypothetical questions are unorthodox, especially for blog posts that are under TOC management. Can you reword into a statemtent with an action?

File: content/en/blog/2020/custom-istio-dns-name/index.md [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+metadata:
+  name: istiocontrolplane-default
+  namespace: istio-operator
+spec:
+  addonComponents:
+    istiocoredns:
+      enabled: true
+{{< /text >}}
+
+When this deploys, we have the `istiocoredns` component that has `coredns` installed as well as the `istio-coredns-plugin` which is used to read `ServiceEntry`s and serve DNS queries for their hostname with IP addresses. The interesting bit is how it's configured out of the box. The default out of the box config is in the `coredns` configmap:
```
Comment (sdake) [inline content/en/blog/2020/custom-istio-dns-name/index.md:61]: to read `ServiceEntry` custom resources

File: content/en/blog/2020/custom-istio-dns-name/index.md [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+        except global
+      }
+
+      prometheus :9153
+      cache 30
+      reload
+    }
+{{< /text >}}
+
+Understanding how [coredns plugins work](https://coredns.io/plugins/grpc/) is helpful here. We see with the `grpc` statement that we're forwarding `global` domains to `127.0.0.1:8053`, and we're forwarding everything else to `/etc/resolv.conf`. The `istio-coredns-plugin` plugin is another container co-deployed with the `coredns` server and is listening on that `127.0.0.1:8053` port within the pod. If we want to change or add domains that can resolve based on what `ServiceEntry`s are created, we need to modify this file.
```
Comment (sdake) [inline content/en/blog/2020/custom-istio-dns-name/index.md:86]: change all references of we to first person.

File: content/en/blog/2020/custom-istio-dns-name/index.md [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+          forward . /etc/resolv.conf {
+            except global
+          }
+          prometheus :9153
+          cache 30
+          reload
+        }
+{{< /text >}}
+
+With this configuration of Istio's `coredns` component, we could have `ServiceEntry`s that have either hostnames of `foo.bar.region` or `foo.bar.global` resolvable by this DNS server.
```
Comment (sdake) [inline content/en/blog/2020/custom-istio-dns-name/index.md:117]: this sentence needs to be reworded into the first person and the usage of ServiceEntry - throughout this blog post - needs to use `ServiceEntry` custom resource - preferably with a link to the SE documentation.

File: content/en/blog/2020/custom-istio-dns-name/index.md [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+
+## Enable stubs for kube-dns
+
+The last step is to add this new `region` domain stub to the `kube-dns` entries. This is covered in the [replicated control-plane docs](/docs/setup/install/multicluster/gateways/#setup-dns).
+
+## Next steps
+
+In this blog post we saw how to customize the domain name used in global service discovery and routing within Istio. There may be other usecases for customizing this DNS resolution behavior, so hopefully this is helpful for others. We encourage you to try out [Istio multi-cluster](/docs/setup/install/multicluster/gateways).
```
Comment (sdake) [inline content/en/blog/2020/custom-istio-dns-name/index.md:125]: can you use periods instead of commas in this sentence :)

Comment (christian-posta) [reply content/en/blog/2020/custom-istio-dns-name/index.md:125]: I can change it around if you feel the sentence reads too long.

File: content/en/blog/2020/custom-istio-dns-name/index.md (hunk 12 above)
Comment (christian-posta) [reply content/en/blog/2020/custom-istio-dns-name/index.md:117]: Im not sure what you mean. "we" is first person. I can update to use ServiceEntry custom resource.

File: content/en/blog/2020/custom-istio-dns-name/index.md (hunk 11 above)
Comment (christian-posta) [reply content/en/blog/2020/custom-istio-dns-name/index.md:86]: "we" is first person?

File: content/en/blog/2020/custom-istio-dns-name/index.md (hunk 10 above)
Comment (christian-posta) [reply content/en/blog/2020/custom-istio-dns-name/index.md:61]: I can change this throughout.

File: content/en/blog/2020/custom-istio-dns-name/index.md (hunk 9 above)
Comment (christian-posta) [reply content/en/blog/2020/custom-istio-dns-name/index.md:43]: I can.

File: content/en/blog/2020/custom-istio-dns-name/index.md (hunk 7 above)
Comment (christian-posta) [reply content/en/blog/2020/custom-istio-dns-name/index.md:12]: > control plane has no dash.

Control plane should have a hyphen when it describes "setup" a noun and is linked ...(https://www.grammarbook.com/punctuation/hyphens.asp) Unless this is some convention used purposefully throughout the docs?

File: content/en/blog/2020/custom-istio-dns-name/index.md (hunk 6 above)
Comment (christian-posta) [reply content/en/blog/2020/custom-istio-dns-name/index.md:10]: Yep, will remove from `.spelling`

File: content/en/blog/2020/custom-istio-dns-name/index.md (hunk 5 above)
Comment (christian-posta) [reply content/en/blog/2020/custom-istio-dns-name/index.md:2]: Im not sure what you're suggesting? Did it fail `make lint`?

File: .spelling (hunk 4 above)
Comment (christian-posta) [reply .spelling:605]: I don't see it https://github.com/istio/istio.io/blob/master/.spelling#L598

**Comment by christian-posta** [comment]
I appreciate the review @sdake .. per your comment, would you recommend this go into a doc and not a blog post?

**Comment by sdake** [comment]
@christian-posta your choice. The blog post is fine if you like since it explains the details of implementation rather than how to use the implementation. You will need to fix the `.spelling` file - that is like a third rail for some reviewers :)

Cheers
-steve

File: content/en/blog/2020/custom-istio-dns-name/index.md (hunk 12 above)
Comment (sdake) [reply content/en/blog/2020/custom-istio-dns-name/index.md:117]: Right. Our docs use the `you` pronoun. I have always steered clear of the `you` pronoun in docs, however, after seeing this method in action in Istio, I understand why You is preferable to `We`.

**Comment by sdake** [comment]
@rcaballeromx or adammil2000 Would you mind having an editor edit this blog post for clarity? I would appreciate it. Once the `.spelling` file is fixed, I'll remove my review on the PR.

File: content/en/blog/2020/custom-istio-dns-name/index.md (hunk 5 above)
Comment (sdake) [reply content/en/blog/2020/custom-istio-dns-name/index.md:2]: A bunch of words were added to the `.spelling`, possibly because you thought they would be caught up in the header spell checking. This is not needed. I will provide further guidance for the `.spelling` file above.

File: .spelling (hunk 4 above)
Comment (sdake) [reply .spelling:605]: Istio uses a default dictionary. The `.spelling` file is an extension to the default dictionary. I am just surprised to see `suffix` was needed to extend the default dictionary. I'd suggest reworking this document to remove all references in `.spelling` by using backticks. If something is required, you can keep it in `.spelling`.

File: content/en/blog/2020/custom-istio-dns-name/index.md (hunk 7 above)
Comment (sdake) [reply content/en/blog/2020/custom-istio-dns-name/index.md:12]: Istio uses this style guide, which I quite like: https://developers.google.com/style/hyphens. I think it well-represents the expectations of the technical writer community. My writing is not pro enough to be considered such a member, but I try :)

File: content/en/blog/2020/custom-istio-dns-name/index.md (hunk 11 above)
Comment (sdake) [reply content/en/blog/2020/custom-istio-dns-name/index.md:86]: I had meant third person. Been a long day - even at 2pm 👍 Please reference this style guideline for followup questions : https://developers.google.com/style/pronouns

File: content/en/blog/2020/custom-istio-dns-name/index.md (hunk 12 above)
Comment (sdake) [reply content/en/blog/2020/custom-istio-dns-name/index.md:117]: https://developers.google.com/style/pronouns

File: content/en/blog/2020/custom-istio-dns-name/index.md (hunk 13 above)
Comment (sdake) [reply content/en/blog/2020/custom-istio-dns-name/index.md:125]: https://developers.google.com/style/pronouns.

File: .spelling [hunk 14]
```diff
@@ -648,6 +656,8 @@ upstreams
 uptime
 URIs
 url
+usecase
```
Comment (sdake) [inline .spelling:659]: this should probably be `use-case`

File: .spelling [hunk 15]
```diff
@@ -648,6 +656,8 @@ upstreams
...
+usecases
```
Comment (sdake) [inline .spelling:660]: and use-cases

File: content/en/blog/2020/custom-istio-dns-name/index.md (hunk 5 above)
Comment (sdake) [reply content/en/blog/2020/custom-istio-dns-name/index.md:2]: I had noticed a bunch of words added to `.spelling`. I believe you may have taken this action to handle spelling errors in the header. I was just indicating the header is not spell-checked and requires extra attention from other reviewers for spelling errors.

Cheers
-steve

File: content/en/blog/2020/custom-istio-dns-name/index.md (hunk 7 above)
Comment (elysdir) [inline content/en/blog/2020/custom-istio-dns-name/index.md:12]: I found the first sentence of this paragraph to be a little bit hard to follow, partly because of the sentence structure and partly because it relies so heavily on multi-syllable words. The sentence is perfectly grammatical, but I wonder if you could make it a little easier to read, maybe by changing it to something like this:

"Multi-cluster Istio lets services communicate with each other, which lets services discover each other."

Or maybe the end of that sentence should be something like "...which allows for service discovery"? I suppose "which enables service discovery" would also be OK in this context.

We usually mildly recommend against using "functionality" and "enable" in this sort of context:

https://developers.google.com/style/word-list#functionality

https://developers.google.com/style/word-list#enable

Those words aren't a big problem, but they do sometimes make docs a little harder to read.

But in this case, "enables" might end up being the best compromise between being brief and being accurate.

Comment (elysdir) [reply content/en/blog/2020/custom-istio-dns-name/index.md:12]: > `you create [Service Entry](link) custom resources in one cluster to reference services running in another cluster.`

Agreed. In particular, it's best to avoid making a keyword into a plural; for more on that, see (https://developers.google.com/style/code-in-text#keywords).

**Comment by elysdir** [comment]
> @rcaballeromx or @adammil2000 Would you mind having an editor edit this blog post for clarity? I would appreciate it.

Just following up on this to mention that I'm now looking over the post. (I'm an editor.) Thanks for asking!

File: content/en/blog/2020/custom-istio-dns-name/index.md (hunk 8 above)
Comment (elysdir) [inline content/en/blog/2020/custom-istio-dns-name/index.md:41]:
```suggestion
This `ServiceEntry` creates an entry for `httpbin.bar.global` in the Istio service-discovery registry. The new entry's endpoints direct traffic to a different cluster's gateway. Without any other configuration changes, this hostname, `httpbin.bar.global`, is not routable. To make it routable, follow the instructions about [configuring the DNS stubbing](/docs/setup/install/multicluster/gateways/#setup-dns) for your Kubernetes cluster. When you set up the DNS stubbing, you tell `kube-dns` to delegate queries for `.global` to the Istio `coredns` component.
```

My proposed revision here is meant to simplify the paragraph, but I may have gone too far and changed the meaning; please ignore or change anything that I got wrong here. My main idea here is to focus on what the reader should do rather than on where the documentation that you're directing them to lives.

(I do like the idea of saying where the docs are, but I think in this particular sentence, it distracts a little bit from the focus of the paragraph. But I may be wrong.)

File: content/en/blog/2020/custom-istio-dns-name/index.md (hunk 9 above)
Comment (elysdir) [inline content/en/blog/2020/custom-istio-dns-name/index.md:43]: Suggested minor changes to the second half of the paragraph:

```suggestion
But what if you don't want to use `.global` as the suffix? What if you have to keep some backward compatibility and the root domain name needs to be something custom? Or what if you want to be able to support multiple root domain suffixes? In the open source project that I work on, [Service Mesh Hub](https://github.com/solo-io/service-mesh-hub), we had this exact scenario. Service Mesh Hub is used to [automate multi-cluster Istio service mesh deployments](https://docs.solo.io/service-mesh-hub/latest/getting_started/), and our users need to route between clusters. The rest of this post explains how to configure routing between clusters in Istio.
```

(The reason I'm removing the hyphen from "open source" is this style guide entry: (https://developers.google.com/style/word-list#open-source).)

File: content/en/blog/2020/custom-istio-dns-name/index.md [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+
+But what if you don't want to use `.global` as the suffix? What if you have to keep some backward compatibility and the root domain name needs to be something custom? Or what if you want to be able to support multiple root domain suffixes? In the open-source project [Service Mesh Hub](https://github.com/solo-io/service-mesh-hub) that I work on, we had this exact scenario. Service Mesh Hub is used to [automate multi-cluster Istio service mesh deployments](https://docs.solo.io/service-mesh-hub/latest/getting_started/), and routing between clusters is a use case our users have. Here's how to configure it in Istio.
+
+## Istio CoreDNS
+
+Istio's operator allows you to deploy the Istio `coredns` addon component.  For example, you can specify the following in the operator config:
```
Comment (elysdir) [inline content/en/blog/2020/custom-istio-dns-name/index.md:47]: Change "addon" to "add-on". (https://developers.google.com/style/word-list#add-on)

Comment (elysdir) [inline content/en/blog/2020/custom-istio-dns-name/index.md:47]: We editors have just been discussing adding the word "config" to our style guide. In general, I think it should either be spelled out as "configuration" or put in code font if it's referring to an entity named `config`. It looks like most of the Istio docs spell it out as "configuration", but I may be misunderstanding.

File: content/en/blog/2020/custom-istio-dns-name/index.md (hunk 10 above)
Comment (elysdir) [inline content/en/blog/2020/custom-istio-dns-name/index.md:61]: Probably best to add a noun after "this" to clarify exactly what it is that's deploying. (https://developers.google.com/style/pronouns#ambiguous-pronoun-references)

File: content/en/blog/2020/custom-istio-dns-name/index.md (hunk 11 above)
Comment (elysdir) [reply content/en/blog/2020/custom-istio-dns-name/index.md:86]: Yeah, best to avoid "we". In addition to the pronouns page that sdake pointed to, see also (https://developers.google.com/style/person).

File: content/en/blog/2020/custom-istio-dns-name/index.md (hunk 12 above)
Comment (elysdir) [inline content/en/blog/2020/custom-istio-dns-name/index.md:117]: I'm not quite sure what this phrase means:

"that have either hostnames of `foo.bar.region` or `foo.bar.global` resolvable by this DNS server."

I think that removing the word "either" would probably make it clearer. Give that a try, and then take another look and see whether it says what you intend it to say.

File: content/en/blog/2020/custom-istio-dns-name/index.md [hunk 17]
```diff
@@ -0,0 +1,125 @@
...
+          cache 30
+          reload
+        }
+{{< /text >}}
+
+With this configuration of Istio's `coredns` component, we could have `ServiceEntry`s that have either hostnames of `foo.bar.region` or `foo.bar.global` resolvable by this DNS server.
+
+## Enable stubs for kube-dns
+
+The last step is to add this new `region` domain stub to the `kube-dns` entries. This is covered in the [replicated control-plane docs](/docs/setup/install/multicluster/gateways/#setup-dns).
```
Comment (elysdir) [inline content/en/blog/2020/custom-istio-dns-name/index.md:121]: Consider changing "This is covered in" to something like "For instructions, see".

File: content/en/blog/2020/custom-istio-dns-name/index.md (hunk 13 above)
Comment (elysdir) [inline content/en/blog/2020/custom-istio-dns-name/index.md:125]:
```suggestion
This blog post explained how to customize the domain name used in global service discovery and routing within Istio. For more information about Istio multi-cluster deployment, see [Replicated control planes](/docs/setup/install/multicluster/gateways).
```

Comment (elysdir) [reply content/en/blog/2020/custom-istio-dns-name/index.md:125]: I don't think you need the line about there being other use cases, nor the "hopefully this is helpful for others" part. But if you do end up keeping those parts, then change "usecases" to "use cases".

File: content/en/blog/2020/custom-istio-dns-name/index.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (sdake) [reply content/en/blog/2020/custom-istio-dns-name/index.md:10]: @elysdir your change SGTM.

File: content/en/blog/2020/custom-istio-dns-name/index.md (hunk 17 above)
Comment (sdake) [reply content/en/blog/2020/custom-istio-dns-name/index.md:121]: nice suggestion, I'll have to start using this!

**Comment by sdake** [comment]
@elysdir pro review. I am really liking that we have real editors, instead of people that barely escaped English 101 like me. 👍 It is a pleasure to virtually meet you!

@christian-posta the release is shipping shortly (t-1 week). Please accept or deny the suggested changes, and certainly fix the spelling file, and we can get this merged prior to the release of Istio 1.6.

Cheers
-steve

**Comment by frankbu** [comment]
@christian-posta Sorry that this PR seems to have slipped through the cracks. Please update the date and make any last minute changes, if you'd like, and we can merge and cherry-pick it right away.

**Comment by frankbu** [comment]
Closing this PR for now.

@christian-posta Please reopen, or send another PR, when it's ready for final review.
